code/retrieval/retriever.py

```diff
@@ -185,82 +190,348 @@ def __init__(self, endpoint_name: Optional[str] = None, query_params: Optional[D
+        except Exception as e:
+            # Any error means the backend doesn't support get_sites or it failed
+            logger.info(f"Backend for endpoint {endpoint_name} does not support get_sites() or it failed: {e}")
+            # Cache None to indicate unsupported
+            self._endpoint_sites_cache[endpoint_name] = None
+            return None
+    
+    async def _endpoint_has_site(self, endpoint_name: str, site: Union[str, List[str]]) -> bool:
+        """
+        Check if an endpoint has data for the requested site(s).
+        
+        If the backend doesn't support get_sites, we assume it might have the site.
+        
+        Args:
+            endpoint_name: Name of the endpoint
+            site: Site name or list of site names to check
+            
+        Returns:
+            True if endpoint has data for at least one of the requested sites
+        """
+        # Handle 'all' case - endpoint should be queried for all sites
+        if site == "all":
+            return True
+            
+        endpoint_sites = await self._get_endpoint_sites(endpoint_name)
+        if endpoint_sites is None:
+            # Backend doesn't support get_sites, assume it might have the site
+            return True
+        elif not endpoint_sites:
+            # Backend supports get_sites but returned empty list
+            return False
+        
+        # Convert site to list for uniform handling
+        if isinstance(site, str):
+            sites_to_check = [site]
+        else:
+            sites_to_check = site
+        
+        # Check if any requested site is available in the endpoint
+        for site_name in sites_to_check:
+            if site_name in endpoint_sites:
+                return True
+        
+        return False
+    
+    def _has_valid_credentials(self, name: str, config) -> bool:
+        """
+        Check if an endpoint has valid credentials based on its database type.
+        
+        Args:
+            name: Endpoint name
+            config: Endpoint configuration
+            
+        Returns:
+            True if endpoint has required credentials
+        """
+        db_type = config.db_type
+        
+        if db_type in ["azure_ai_search", "snowflake_cortex_search"]:
+            # These require API key and endpoint
+            return bool(config.api_key and config.api_endpoint)
+        elif db_type == "opensearch":
+            # OpenSearch requires endpoint and credentials (api_key)
+            return bool(config.api_endpoint and config.api_key)
+        elif db_type == "qdrant":
+            # Qdrant can use either local path or remote URL
+            if config.database_path:
+                return True  # Local file-based storage
+            else:
+                return bool(config.api_endpoint)  # Remote server (api_key is optional)
+        elif db_type == "milvus":
+            # Milvus requires database path
+            return bool(config.database_path)
+        else:
```

Comment:
I don't think this is right.  Milvus is another endpoint/API key combo, not a database path, in the main branch currently.